man/fi_profile.3.md

```diff
@@ -0,0 +1,281 @@
+interface. Providers define events and use them to report when an operation 
+has occured or a state changed where one or more variables may change their 
+values. Users can register callback functions to one or more events, and 
+receive notifications from providers once an event occurs.
+
+In order to support the profiling interface, an endpoing or a domain in a 
+provider needs to maintain a fid_profile object, and implement the profiling 
+operations there. To start a profiling, an open call must be made on the 
+target endpoint or domain object to obtain the fid_profile object.  If the 
+target object does not support profiling, the open call will fail with error 
+code -FI_ENOSYS. Once a fid_profile object is obtained successfully, it can be 
+used to query available variables or events, and read variables' values.
+
+# VARIABLES
+
+The profiling variables are defined in providers and are accessible through 
+the profiling interface. Variables are specified using struct 
+fi_profile_desc described below. A variable can be a configuration setting, 
+a runtime state or a counter associated with an endpoint or a domain. 
+Some variables are common and available in most providers, while others are 
+provider-specific, and are only available in one provider. Users can query 
+available variables and their values using the profiling interface.
+
+Provider specific variables are outside the scope of this man page. libfabric 
+defined variables and their datatypes are described below.
+
+## FI_VAR_UNEXP_MSG_CNT (uint64_t)
+
+This variable returns the current number of unexpected messages that have been
+received by the provider which are waiting for matching receive buffers.
+
+# EVENTS
+
+Profiling events notify users that an operation has occurred or the state of
+an object has changed. Events typically indicate that the values of one or 
+more variables have been modified. Similar to variables, events may be 
+pre-defined by libfabric or be provider specific. Users can register a 
+callback to be notified when an event occurs.  libfabric defined events 
+and the data that corresponds with the event, if any, are described below.
+
+Applications may read variable values from the event callback, but may not 
+register or unregister for events, query for profiling variable or events, 
+or call the profile lock and unlock functions
+
+libfarbic pre-defined events include:
+
+## FI_EVENT_UNEXP_MSG_RECVD 
+
+This event indicates that an unexpected message was received and has been
+queued by the provider. 
+
+## FI_EVENT_UNEXP_MSG_MATCHED
+
+This event indicates that an unexpected messages that was previously queued
+by the provider has been matched. 
+
+These two events can notify a user without any data and rely on the user 
+to read variables (e.g, FI_VAR_UNEXP_MSG_CNT) when needed.  It also
+can notify user with a fi_cq_tagged_entry data ("param") which contains 
+<src_addr, tag, size> to indicate how the event occured.
+
```

Comment:
Is it a hole, that you can get the current number of unexpected messages, but not get information on any previously received unexpected messages (at time of opening the profiling) until those messages are matched with a posted receive? Is the intent for profiling be started before the EP is enabled?

Reply:
With the current pre-defined variable (unexpected message count) and events, we cannot get info on previously received unexpected messages.  This can be added as a provider-specific variable, say "unexpected_info_queue",  a provider needs to maintain this "queue",  and the user can use the profiling interface to query this variable value, to get the info in the queue.   We are looking for input/feedback to add more meaningful variables and events.  Profiling is enabled on EP after PE is created.

Reply:
The CXI provider currently has a provider specific call that allows the number of unexpected messages to be queried and for retrieving the existing entries. They are returned in an array of CQ entries and an optional array of addresses (so no new cq format). The unexpected list itself is maintained by hardware. Ideally as you indicate, it would be great if this "pull" model could fit into the profile API. This work was done to provide CXI prototype support for unexpected messages in the MPI Message Queue Dumping Interface.

Reply:
@swelch - See my comment above.  We looked at using one of the existing cq formats, but those lack the ability to return a src_addr, which may be useful.  So I think we will end up with a new structure.  I think we work to get the pull model you're describing to work, which will help prove that the API is adaptable.  The pull model should be the same as reading a variable -- just a more complex one -- that can return an array of structs (like the one defined above).